onnxruntime/test/onnx/callables.h

```diff
@@ -0,0 +1,150 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#pragma once
+
+#include <type_traits>
+#include <tuple>
+
+namespace onnxruntime {
+namespace test {
+
+/// <summary>
+/// This class serves as a callable entity that can be used
+/// either in a threadpool or as a callback fro async operations. It supports both
+/// plain functions as well as member functions and provides type
+/// erasure functionality. Unlike std::function, it
+/// does do any memory allocations and it is cheap to copy.
+/// </summary>
+/// 
+/// <example>
+/// Plain function:
+///   int g(void* context, float);
+///   Callable<int, float> cb(nullptr, g);
+///   int result = cb.Invoke(1.2f);
+///
+/// Class method:
+/// class A {
+///    int f(float);
+/// };
+/// 
+/// A a;
+/// CallableFactory<A, int, float> f(&a);
+/// - We get back Callable<int, float> as above but call instance method
+/// auto cb = f.GetCallable<&A::f>();
+/// int result = cb.Invoke(1.2f);
+/// </example>
+/// 
+/// <typeparam name="Result">Call back result type</typeparam>
+/// <typeparam name="...Args">Types of args for the callable</typeparam>
+template <typename    Result,
+          typename... Args>
+class Callable {
+public:
+
+  using
+  ContextType = void;
+
+  using
+  ResultType = Result;
+
+  using
+  FunctionType = ResultType(*)(ContextType*, Args...);
+
+  // You can use this directly if you have a simple
+  // function like C-function or a stateless lambda
+  // For methods we want to make use of AsyncMethodCallback
+  // And covert it to a Callable
+  Callable(ContextType*  context,
+    FunctionType func) :
+    context_(context),
+    func_(func) {
+  }
+
+  Callable() : Callable(nullptr, nullptr) {}
+
+  Callable(const Callable&) = default;
+  Callable& operator=(const Callable&) = default;
+
+  Callable(Callable&&) = default;
+  Callable& operator=(Callable&&) = default;
+
+  void Clear() {
+    context_ = nullptr;
+    func_ = nullptr;
+  }
+
+  operator bool() const { return func_ != nullptr; }
+
+  bool Valid() const { return func_ != nullptr; }
+
+  ContextType* GetContext() const { return context_; }
+
+  FunctionType GetFunctionPtr() const { return func_; }
+
+  // Differentiate instantiation on the presence
+  // of return value. The below two overloads
+  // must be mutually exclusive
+  template<typename T = ResultType>
+  typename std::enable_if<!std::is_same<void, T>::value, T>::type
```

Comment:
>std::enable_if<!std::is_same<void, T>::value, T>::type [](start = 11, length = 54)

enable_if_t for simplicity?

Reply:
_t is not supported by all of our compilers.

---
In reply to: [491078654](https://github.com/microsoft/onnxruntime/pull/5169#discussion_r491078654) [](ancestors = 491078654)